system/include/emscripten/val.h

```diff
@@ -65,6 +65,8 @@ namespace emscripten {
             EM_VAL _emval_get_property(EM_VAL object, EM_VAL key);
             void _emval_set_property(EM_VAL object, EM_VAL key, EM_VAL value);
             EM_GENERIC_WIRE_TYPE _emval_as(EM_VAL value, TYPEID returnType, EM_DESTRUCTORS* destructors);
+            int64_t _emval_as_int64(EM_VAL value, TYPEID returnType);
+            uint64_t _emval_as_uint64(EM_VAL value, TYPEID returnType);
```

Comment:
I am not sure about this and would like feedback. The issue was that `emscripten::val::as<int64_t>()` was failing because emscripten was converting the values to `int64_t` and that was conflicting with `EM_GENERIC_WIRE_TYPE` which is a `double`.

I guess the other option would be to pack the `int64_t` into a `double` over the wire, but that seems a shame when emscripten handles the `bigint->int64_t` conversion so nicely.

Reply:
I guess the other option would be to expand `EM_GENERIC_WIRE_TYPE` into an enum?

Reply:
What does "conflict" mean here?

Packing an i64 as a double is risky, as some VMs canonicalize doubles.

Reply:
I was getting an exception when `emscripten::val::as<int64_t>()` returned a `BigInt` on the JS side. I think that the error came from trying to pack the `BigInt` into the `EM_GENERIC_WIRE_TYPE`. I will dig up the exception. As far as I could tell the exception was coming from the VM.

Reply:
Here is the error I was getting before I added the specialization:
```
test:1246 exception thrown: TypeError: Cannot convert a BigInt value to a number,TypeError: Cannot convert a BigInt value to a number
    at unsigned long long emscripten::val::as<unsigned long long>() const (http://localhost:5000/test.wasm:wasm-function[32]:0x1347)
    at __original_main (http://localhost:5000/test.wasm:wasm-function[26]:0x104a)
    at main (http://localhost:5000/test.wasm:wasm-function[54]:0x19bb)
    at http://localhost:5000/test.js:1503:22
    at callMain (http://localhost:5000/test.js:5805:15)
    at doRun (http://localhost:5000/test.js:5875:23)
    at http://localhost:5000/test.js:5886:7
```

So, when the [JS implementation of `_emval_as`](https://github.com/cuberoot/emscripten/blob/main/src/embind/emval.js#L282) returns a `BigInt` from the `toWireType` call [here](https://github.com/cuberoot/emscripten/blob/cuberoot/embind-int64/src/embind/embind.js#L613-L620), it chokes with the above error.

Reply:
@kripken Any thoughts ☝️

Reply:
I see, makes sense. Yes, we need to avoid casting to a double, and we need to have an explicit BigInt cast. So a specialization sounds right.